pkg/signal/signal.go

```diff
@@ -6,180 +6,286 @@ package signal
+	var set signalSet
+	for _, signal := range signals {
+		if signal >= SignalTypeMax {
+			return fmt.Errorf("Signal number not supported: %d", signal)
+		}
+		set |= signalSet(1) << signal
 	}
+	sm.setUnmuted(set)
 	return nil
 }
 
-// RegisterChannel registers a go channel for a given signal.
-func RegisterChannel(signal int, ch chan<- SignalData) error {
-	if signal >= SignalChanInvalid {
-		return fmt.Errorf("Signal number not supported: %d", signal)
+// RegisterChannel registers a go channel for the given signals.
+// 'sm' is not a receiver so that this can be a generic function.
+func RegisterChannel[T fmt.Stringer](sm SignalRegistrar, ch chan<- T, signals ...SignalType) error {
+	if ch == nil {
+		return fmt.Errorf("Nil channel")
 	}
-	if channels[signal] != nil {
-		return fmt.Errorf("Channel for signal number already registered: %d", signal)
+
+	handler := func(reader io.Reader) (string, error) {
+		if reader == nil {
+			close(ch)
+			return "", io.EOF
+		}
+		var data T
+		if err := binary.Read(reader, byteorder.Native, &data); err != nil {
+			return "", err
+		}
+		select {
+		case ch <- data:
+		default:
+			return "", ErrFullChannel
+		}
+		return data.String(), nil
 	}
 
-	channels[signal] = ch
-	return nil
+	return sm.RegisterHandler(handler, signals...)
 }
 
-// SetupSignalListener bootstraps signal listener infrastructure.
-func SetupSignalListener() {
-	once.Do(func() {
-		var err error
+// registerHandler registers a signal handler for the given signals.
+func (sm *signalManager) RegisterHandler(handler SignalHandler, signals ...SignalType) error {
+	if sm.events != nil {
+		return fmt.Errorf("Runtime registration not supported")
+	}
 
-		path := oldBPF.MapPath(signalmap.MapName)
-		signalMap, err := ebpf.LoadPinnedMap(path, nil)
-		if err != nil {
-			log.WithError(err).Warningf("Failed to open signals map")
-			return
+	for _, signal := range signals {
+		if signal >= SignalTypeMax {
+			return fmt.Errorf("Signal number not supported: %d", signal)
 		}
-		events, err = perf.NewReader(signalMap, os.Getpagesize())
-		if err != nil {
-			log.WithError(err).Warningf("Cannot open %s map! Ignoring signals!",
-				signalmap.MapName)
-			return
+		if sm.handlers[signal] != nil {
+			return fmt.Errorf("Channel for signal number already registered: %d", signal)
 		}
+	}
 
-		go func() {
-			log.Info("Datapath signal listener running")
-			for {
-				record, err := events.Read()
-				switch {
-				case err != nil:
-					signalCollectMetrics(nil, "error")
-					log.WithError(err).WithFields(logrus.Fields{
-						logfields.BPFMapName: signalmap.MapName,
-					}).Errorf("failed to read event")
-				case record.LostSamples > 0:
-					signalCollectMetrics(nil, "lost")
-				default:
-					signalReceive(&record)
+	for _, signal := range signals {
+		sm.handlers[signal] = handler
+		sm.setUnmuted(signalSet(1) << signal)
+	}
+	return nil
+}
+
+// open the signalmap and start signal listener. Called after all the handlers have registered.
```

Comment:
the signal map itself should already be open -> maybe change the description and/or change method name

Reply:
Right, renamed these to `start()`/`stop()` and adjusted comments accordingly.